### ISSUE TITLE
Container can't be shrunk past children when clip content is enabled

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot#38575]
**Godot version:**
3.2.1

**OS/device including version:**
Windows 10 Pro Education 64 bit

**Issue description:**
Containers like VBoxContainer can't be shunk down to clip its content when clip content is enabled

this makes the clip content option useless on containers since there will never be anything to clip

**Steps to reproduce:**
- Add a VBoxContainer
- Add Two buttons to it
- Enable Clip content on the VBoxContainer
- Try to shrink the VBoxContainer to only show 1 button

**Minimal reproduction project:**
[Container Clip Bug.zip](https://github.com/godotengine/godot/files/4600926/Container.Clip.Bug.zip)

**Comment by groud** [comment]
Maybe we could allow for it, but a container that does not set boundaries to its children kind of defeat the purpose of a container. That could be discussed though.

**Comment by Shadowblitz16** [comment]
it would still define the boundaries the only difference is that the visual/interactive boundaries would be different then the content boundaries

**Comment by follower** [comment]
> Maybe we could allow for it, but a container that does not set boundaries to its children kind of defeat the purpose of a container.

To clarify, if I'm understanding correctly, this seems to be the *opposite* of what @Shadowblitz16 is describing/requesting.

If I understand correctly, @Shadowblitz16 is saying the expectation is that "clip content" set on the container *should* prevent from displaying any part of the child content extending past the container's boundary but that doesn't match the behaviour seen.

e.g. Expected behaviour when clip content is enabled:

```
|Container|
|ChildLong|
```

vs Observed behaviour when clip content is enabled:

```
|Container     |
|ChildLongLabel|
```

Please correct me if I've incorrectly interpreted of what either of you is describing. :)

Based on my current understanding of the situation I've investigated & think I know what's happening. Will follow-up in a separate comment.

**Comment by follower** [comment]
**TL;DR: Put the `VBoxContainer` in a `Control` with clipping enabled.**

### Context

Based on my own experiences I've learned that Godot's container layout system is very powerful but that there are sometimes unexpected interactions between controls & unexpected influence from properties. (In addition to general bug/documentation issues.)

### One (partial) approach

In the situation @Shadowblitz16 describes (specifically for buttons) the short answer is:

 * To enable horizontal clipping the `clip_text` property on the button *also* needs to be enabled *but* even then you won't be able to make the container narrower than the button's margin/padding. This also won't help with vertical clipping.

    (It's possible to see the minimum size a control will take by unchecking all its `size_flags_*` values.)

 * Actually, there's a better option, see below.

### Lessons from the docs: `BoxContainers` are about minimum sizes

In situations like this I find it helpful to read the tool tips/docs for properties and indeed I learned a couple of important points (and even re-re-reading just now :D ) (quote emphasis mine):

 * From [`BoxContainer`](https://docs.godotengine.org/en/3.2/classes/class_boxcontainer.html#class-boxcontainer) (parent/base class of `HBoxContainer`, `VBoxContainer`):
   > *Arranges child controls vertically or horizontally, and rearranges the controls automatically **when their minimum size changes**.*

 * From [`Control.rect_min_size`](https://docs.godotengine.org/en/3.2/classes/class_control.html#class-control-property-rect-clip-content):
   > ***If** you **set it to a value greater than (0, 0)**, the node’s bounding rectangle will always **have at least this size, even if its content is smaller**. If it’s **set to (0, 0), the node sizes automatically to fit its content**, be it a texture or child nodes.*

So, from this, I think that it's best to think about the box container classes as being explicitly *related to the **minimum size** of the controls contained within* (and implicitly *not* **smaller** than minimum size. )

### Minimum sizes are easily influenced from many places :)

In addition, be aware that a Control's minimum size can be influenced from some unexpected places. e.g. I recently discovered that `LineEdit`'s minimum size is in part defined by a *theme* property [`minimum_spaces`](https://docs.godotengine.org/en/3.2/classes/class_lineedit.html#theme-properties) that defaults to a *minimum width of 12 characters* even if, say, `max_length` property is lower. (I guess the source code of each control is the definitive arbiter of what gets to influence it.)

### Use a plain `Control` node to get clipping of `BoxContainers`

So, how can you get the clipping result you want?

From a quick test, if you put the box container inside of a bare `Control` node (and enable `rect_clip_content`) then clipping should occur--based on the size of the `Control` node.

Now, whether that will be automatic enough for your purposes may require further investigation--at this point I stopped going down this particular rabbit/yak hole. :D

**Comment by Shadowblitz16** [comment]
so what I am trying to say is.. 

I think we should be able to shrink containers down past their content boundaries if clip content is enabled for the container.

this would not effect the size of the contents inside it would just clip it visually and wouldn't let the mouse interact with it.

![image](https://user-images.githubusercontent.com/2453707/81511554-15367f00-92cf-11ea-9cae-e1c098a8e473.png)

**Comment by zerotrickpony** [comment]
Follower's workaround was very helpful thank you! I fixed my own issue by inserting a single-child Container in the hierarchy. (Inserting a ScrollContainer also did more like what I expected, except that I don't want the scrollbars.)

I found this thread because it doesn't seem like `clip_contents` does anything useful on VBoxContainer? I am a godot novice but I have decades of experience with HTML5 layout. I expected nesting VBoxContainers to behave like `display:flex`, and that `clip_contents=true` should behave like `overflow:hidden` on a child. 

### What I observe (v4.6)
Suppose I want to clip a large child in a nested layout:

- vbox1 expand=true desired_size_from_parent.y=500
  - vbox2 clip_contents=true expand=true
    - very_big_child3 min_size.y=1000
  - somelabel4 size.y=50

This causes vbox1 and vbox2 to both ignore their parent's constraints and size themselves to the large inner child:
- vbox1 min_size.y=1050  // unwanted overspill of parent
  - vbox2 min_size.y=1000  // unwanted oversize, despite clip_contents
    - very_big_child3 size.y=1000  // draws fully
  - somelabel4 size.y=50  // pushed off the bottom of the screen by the large child above

### What I expect
Instead, what I want is to have `vbox2.clip_contents=true` to cause its minumum size to remain zero, allowing vbox1 to inherit its desired parent size instead of overspilling its parent. Like so:

- vbox1 min_size.y=0 size.y=500 // sized from parent
  - vbox2 min_size.y=0 size.y=450 // expanded to share space with sibling, but no more
    - very_big_child3 // y=450 drawn, and then clipped by the parent so the remaining 550 is hidden
  - somelabel4 size.y=50 // not pushed out of the layout

...this behavior is what ScrollContainer does when its scrollbar mode is `ALWAYS`, so I'd want VBoxContainer to do something similar.
